src/sage/groups/abelian_gps/abelian_group.py

```
@@ -573,19 +573,56 @@ def is_isomorphic(left, right):
 
         INPUT:
 
-        - ``right`` -- anything
+        - ``right`` -- a group
+
+        OUTPUT: boolean; whether ``left`` and ``right`` are isomorphic as groups
+
+        .. WARNING::
 
-        OUTPUT: boolean; whether ``left`` and ``right`` are isomorphic as abelian groups
+           When ``right`` is not an instance of ``AbelianGroup`` and both ``left`` and ``right``
+           are infinite, this method will throw a ``GAPError``.
 
         EXAMPLES::
 
             sage: G1 = AbelianGroup([2,3,4,5])
             sage: G2 = AbelianGroup([2,3,4,5,1])
             sage: G1.is_isomorphic(G2)
             True
+
+        TESTS:
+
+        Check that :issue:`39893` is fixed::
+
+            sage: G = AbelianGroup([2, 3])
+            sage: H = G.permutation_group()
+            sage: G.is_isomorphic(H)
+            True
+            sage: G = AbelianGroup([2, 3])
+            sage: H = PermutationGroup([(1, 2, 3), (4, 5)])
+            sage: G.is_isomorphic(H)
+            True
+            sage: G = AbelianGroup([2, 3, 0])
+            sage: H = PermutationGroup([(1, 2, 3), (4, 5)])
+            sage: G.is_isomorphic(H)
+            False
+            sage: G = AbelianGroup([0])
+            sage: H = FreeGroup(1)
+            sage: G.is_isomorphic(H)
+            Traceback (most recent call last):
+            ...
+            sage.libs.gap.util.GAPError: Error, cannot test isomorphism of infinite groups
+            sage: G = AbelianGroup([2, 3, 0])
+            sage: H = FreeGroup(1)
+            sage: G.is_isomorphic(H)
+            Traceback (most recent call last):
+            ...
+            sage.libs.gap.util.GAPError: Error, cannot test isomorphism of infinite groups
```

I suggest catch it and re-raise NotImplementedError. Avoid exposing too much implementation detail, especially since we don't want upstream catch `GAPError`.

---

Good point. I did see some other parts of the library do throw a `GAPError`, e.g. the `PrimitiveGroups` function from `permgroups_named.py` throws a `GAPError` when called with too large a degree:
```sage
sage: PrimitiveGroups(2^12).cardinality()
Traceback (most recent call last):
...
GAPError: Error, Primitive groups of degree 4096 are not known!
```
But I see the utility of making the error message more user friendly. I could wrap the isomorphism check in a try-except block as follows:
```python3
try:
    iso = left._libgap_().IsomorphismGroups(right)
except GAPError as e:
    raise NotImplementedError(e) from None
```
This would catch any `GAPError` and retype it as a `NotImplementedError` hiding the original exception context. The only issue is that this would catch all `GAPErrors` indiscriminately, which might cause weird and confusing error messages if there are `GAPErrors` unrelated to isomorphisms checking of infinite groups. What are your thoughts on this?

---

that's reasonable, catch and reraise is slower too.

Maybe then just modify the documentation to not guarantee that GAPError will definitely be raised.

(some code might rely on that anyway, they really shouldn't)

---

I ended up just removing the reference to `GAPError` in the docs and keeping the code as-is. Catching and reraising didn't seem like the correct solution here.